packages/async-flow/src/async-flow.js

```diff
@@ -0,0 +1,417 @@
+    valueShape: M.error(),
+  });
+
+  const eagerWakers = outerZone.setStore(`asyncFuncEagerWakers`, {
+    keyShape: M.remotable('flow'),
+  });
+
+  const tmp = makeEphemera(() => ({
+    membrane: /** @type {ReplayMembrane} */ (
+      /** @type {unknown} */ (undefined)
+    ), // initialized by restart
+  }));
+
+  /**
+   * So we can give out wrapper functions easily and recover flow objects
+   * for their activations later.
+   */
+  const flowForOutcomeVowKey = outerZone.mapStore('flowForOutcomeVow', {
+    keyShape: M.remotable('vowishKey'),
+    valueShape: M.remotable('asyncFlow'),
+  });
+
+  /**
+   * @param {Zone} zone
+   * @param {string} tag
+   * @param {GuestAsyncFunc} [optGuestAsyncFunc]
+   * @param {{ startEager?: boolean }} [options]
+   */
+  const prepareAsyncFlowKit = (
+    zone,
+    tag,
+    optGuestAsyncFunc = undefined,
+    options = {},
+  ) => {
+    const {
+      // May change default to false, once instances reliably wake up
+      startEager = true,
+    } = options;
+
+    const internalMakeAsyncFlowKit = zone.exoClassKit(
+      tag,
+      AsyncFlowIKit,
+      (activationThis, activationArgs) => {
+        harden(activationThis);
+        harden(activationArgs);
+        const log = makeLogStore();
+        const bijection = makeWeakBijection();
+
+        return {
+          activationThis, // replay starts by reactivating with these
+          activationArgs, // replay starts by reactivating with these
+          log, // log to be accumulated or replayed
+          bijection, // membrane's guest-host mapping
+          outcomeKit: makeVowKit(), // outcome of activation as host vow
+          isDone: false, // persistently done
+        };
+      },
+      {
+        flow: {
+          restart(func = optGuestAsyncFunc, eager = startEager) {
+            if (func === undefined) {
+              throw Fail`Function must either be in prepareAsyncFlowKit or restart args: ${q(
+                tag,
+              )}`;
+            }
+            const { state, facets } = this;
+            const {
+              activationThis,
+              activationArgs,
+              log,
+              bijection,
+              outcomeKit,
+            } = state;
+            const { flow, admin, wakeWatcher } = facets;
+
+            if (failures.has(flow)) {
+              failures.delete(flow);
+            }
+
+            !state.isDone ||
+              // separate line so I can set a breakpoint
+              Fail`Cannot restart a done flow ${flow}`;
+
+            admin.reset();
+            if (eager) {
+              eagerWakers.add(flow);
+            }
+
+            const wakeWatch = vowish => {
+              // Extra paranoid because we're getting
+              // "promise watcher must be a virtual object"
+              // in the general vicinity.
+              zone.isStorable(vowish) ||
+                Fail`vowish must be storable in this zone (usually, must be durable): ${vowish}`;
```

Comment:
As discussed, a promise cannot be stored durably, even if it is imported / externally decided. While I expect that we'll deal mostly with host promises that resolve "immediately" to a vow, we can't assume that. As such we should "wrap" any host promises into a locally created vow, which is resolved by using `watch` on the host promise.

Reply:
Per our slack conversation (which we should capture), I will now insist on only vow on the host side, rejecting promises rather than autocoercing them. Seems less user friendly until we get to the diagnostics needed around perverse promise cases. Better to just leave this coercion for the users to do manually. At least as if this first PR. Resolving.